### ISSUE TITLE
CombineMeshes crash

### ISSUE THREAD

**Issue by BlocInBloc** [atteneder/glTFast#100]
Hi !

Unity Editor (2019.4.13f1) crash when we call  "mesh.CombineMeshes(CombineInstance[])" on imported objects with GltFast.

Unity editor logs 
```
Obtained 9 stack frames.
#0  0x00000106bec4aa in Mesh::RecalculateSubmeshBoundsInternal(unsigned int)
#1  0x00000106bec76c in Mesh::RecalculateSubmeshBounds(unsigned int, MeshUpdateFlags)
#2  0x00000106c2095f in CombineMeshIndicesForStaticBatching(std::__1::vector<CombineInstance, std::__1::allocator<CombineInstance> > const&, Mesh&, bool, bool)
#3  0x00000106c24da5 in CombineMeshes(std::__1::vector<CombineInstance, std::__1::allocator<CombineInstance> > const&, Mesh&, bool, bool)
#4  0x00000106c26b74 in MeshScripting::CombineMeshes(Mesh*, dynamic_array<MeshScripting::MeshCombineInstance, 0ul> const&, bool, bool, bool)
#5  0x000001075ca585 in Mesh_CUSTOM_CombineMeshesImpl(ScriptingBackendNativeObjectPtrOpaque*, ScriptingBackendNativeArrayPtrOpaque*, unsigned char, unsigned char, unsigned char)
#6  0x0000017eeadf29 in  (wrapper managed-to-native) UnityEngine.Mesh:CombineMeshesImpl (UnityEngine.Mesh,UnityEngine.CombineInstance[],bool,bool,bool) {0x7fe825772120} + 0xf9 (0x17eeade30 0x17eeadfe5) [0x177031c80 - Unity Child Domain]
#7  0x007fe82d4ed890 in (Unknown)
#8  0x007ffeebdc96a0 in (Unknown)
```

The script to reproduce the crash 
```
public class CombineTest : MonoBehaviour {
    public MeshFilter obj1;
    public MeshFilter obj2;

    [ContextMenu ("Combine")]
    public void Combine () {
        List<Tuple<Mesh, Transform>> datas = new List<Tuple<Mesh, Transform>> ();
        datas.Add (new Tuple<Mesh, Transform> (obj1.sharedMesh, obj1.transform));
        datas.Add (new Tuple<Mesh, Transform> (obj2.sharedMesh, obj2.transform));
        Mesh combinedMesh = Combine (datas);
    }

    public static Mesh Combine (List<Tuple<Mesh, Transform>> datas) {
        CombineInstance[] combine = new CombineInstance[datas.Count];

        for (int indexMeshFilter = 0; indexMeshFilter < datas.Count; indexMeshFilter++) {
            combine[indexMeshFilter].mesh = datas[indexMeshFilter].Item1;
            combine[indexMeshFilter].transform = datas[indexMeshFilter].Item2.localToWorldMatrix;
        }

        Mesh mesh = new Mesh ();
        mesh.CombineMeshes (combine);
        return mesh;
    }
}
```

**Comment by atteneder** [comment]
Sorry for the crash and the late reply.

This may be related to #86 
I'd have to test, but feel free to try it right away.

**Comment by BlocInBloc** [comment]
We tried to force
`mesh.UploadMeshData(false);`
it still crashes with same editor logs.

**Comment by BlocInBloc** [comment]
We found a solution, we loop over the submesh in our combine function.
It doesn't crash anymore and it works.
Thank you for your help !

**Comment by JSWorkspace1** [comment]
@BlocInBloc can you elaborate on your solution or post a code snippet? I have the same problem since i updated to version 2.2

**Comment by atteneder** [comment]
> @BlocInBloc can you elaborate on your solution or post a code snippet? I have the same problem since i updated to version 2.2

Have you gone through [Readable Mesh Data](https://github.com/atteneder/glTFast/blob/main/Documentation%7E/glTFast.md#readable-mesh-data) in the docs?

**Comment by JSWorkspace1** [comment]
Yes i even debugged the #if GLTFAST_KEEP_MESH_DATA  part and it definitely goes in there but it still crashes on    mesh.CombineMeshes(combine);

**Comment by atteneder** [comment]
Sorry to hear that. I'll reopen the issue and look into it.

Can you provide a test file plus the code snippet that fails?

**Comment by JSWorkspace1** [comment]
code is basically the same as  @BlocInBloc   so i am interested in his solution. 
```
    foreach (var modelKeyPair in Models)
            {          
                var meshFilters = modelKeyPair.Value.GetComponentsInChildren<MeshFilter>();
                var combine = new CombineInstance[meshFilters.Length];

                var i = 0;
                while (i < meshFilters.Length)
                {
                    combine[i].mesh = meshFilters[i].sharedMesh;          
                    combine[i].transform = meshFilters[i].transform.localToWorldMatrix;
                    i++;
                }
             
                var mesh = new Mesh();
                mesh.CombineMeshes(combine); //crashes here 


             ... 
            }
       

```   

Here the full Class and an example gltf Object that i used 

[FileAndCode.zip](https://github.com/atteneder/glTFast/files/5591436/FileAndCode.zip)

**Comment by LoicFr** [comment]
Hello,

This issue is still there in the release 3.0.2.
Has anyone found a workaround for this ?

**Comment by Blocovin** [comment]
Hi,

Sorry for the delay, here's how we got around the crash:

```cs
public static Mesh Combine (List<Tuple<Mesh, Transform>> datas) {
    List<CombineInstance> combine = new List<CombineInstance> (datas.Count);

    for (int indexMeshFilter = 0; indexMeshFilter < datas.Count; indexMeshFilter++) {
        Mesh tmpMesh = datas[indexMeshFilter].Item1;
        if (tmpMesh.subMeshCount > 0) {
            for (int index = 0; index < tmpMesh.subMeshCount; index++) {
                CombineInstance combineInstance = new CombineInstance ();
                combineInstance.mesh = tmpMesh.GetSubmesh (index);
                combineInstance.transform = datas[indexMeshFilter].Item2.localToWorldMatrix;
                combine.Add (combineInstance);
            }
        } else {
            CombineInstance combineInstance = new CombineInstance ();
            combineInstance.mesh = tmpMesh;
            combineInstance.transform = datas[indexMeshFilter].Item2.localToWorldMatrix;
            combine.Add (combineInstance);
        }
    }

    Mesh mesh = new Mesh ();
    mesh.CombineMeshes (combine.ToArray (), false, true);
    return mesh;
}
```
Hope it help !

**Comment by LoicFr** [comment]
Hello

Thank you for your quick reply.

I don't understand your code because:
- `if (tmpMesh.subMeshCount > 0)` is (almost) always true (otherwise the mesh is empty so it doesn't need to be combined);
- `tmpMesh.GetSubmesh (index)` doesn't exist. Is it a typo ? GetSubMesh exists but it doesn't return a Mesh. I think you have implemented your own GetSubmesh function which returns a Mesh and this is this function which prevents the bug to appear. Is it possible to share it please ?

If I use CombineMeshes only on meshes with exactly one subMesh, it still crashes, so it seems to be not linked with the number of subMeshes. Moreover the Unity Editor crashes when selecting a LODGroup which uses meshes imported with GLTFast. Meshes generated by GLTFast seem to be not fully supported by Unity functions...

**Comment by Blocovin** [comment]
> if (tmpMesh.subMeshCount > 0) is (almost) always true (otherwise the mesh is empty so it doesn't need to be combined);

Indeed you are right.

> tmpMesh.GetSubmesh (index) doesn't exist.

I forgot to share you our implementation of GetSubmesh

```cs
public static Mesh GetSubmesh (this Mesh aMesh, int aSubMeshIndex) {
    if (aSubMeshIndex < 0 || aSubMeshIndex >= aMesh.subMeshCount){
        return null;
    }
    int[] indices = aMesh.GetTriangles (aSubMeshIndex);
    Vertices source = new Vertices (aMesh);
    Vertices dest = new Vertices ();
    Dictionary<int, int> map = new Dictionary<int, int> ();
    int[] newIndices = new int[indices.Length];
    for (int i = 0; i < indices.Length; i++) {
        int o = indices[i];
        int n;
        if (!map.TryGetValue (o, out n)) {
            n = dest.Add (source, o);
            map.Add (o, n);
        }
        newIndices[i] = n;
    }
    Mesh m = new Mesh ();
    dest.AssignTo (m);
    m.triangles = newIndices;
    return m;
}
```

**Comment by LoicFr** [comment]
Thank you very much it works now ! And it is not necessary to define `GLTFAST_KEEP_MESH_DATA`.

It is needed to add these lines at the beginning of `public void AssignTo(Mesh aTarget)` to support meshes with more than 65 536 vertices:
```
if (verts.Count > 65535)
     aTarget.indexFormat = UnityEngine.Rendering.IndexFormat.UInt32;
```

Globally it works but this is not a very optimised workaround so I hope it will be fixed directly in GLTFast in a future release.

**Comment by atteneder** [comment]
Hi all,

Thanks for being so patient and providing awesome workarounds here!

TL;DR: Change the variable `flags` in `PrimitiveCreateContext.cs` line 55 to:

```csharp
MeshUpdateFlags flags = MeshUpdateFlags.DontNotifyMeshUsers | MeshUpdateFlags.DontResetBoneBounds | MeshUpdateFlags.DontValidateIndices;
```

As you can see, I tried to speed up mesh creation by avoiding the recalculation of the bounding box (`MeshUpdateFlags.DontRecalculateBounds`). Works, as long as you don't use it in Combine, I guess.

This obviously is a Unity bug. Even if glTFast does something wrong, a proper error message should be the result instead of a crash.

@Blocovin : You script sails around the problem really nicely, but I presume fetching the mesh data (especially via the simple, managed interface) will not be performant.

I have to figure out, which solution is best:

- Let Unity recalculate the bounds (like the solution above)
- Manually calculate bounds via Jobs (maybe add to existing jobs)
- Set dummy bounds (just to avoid the crash...probably open up other issues)
- Retrieve bounds from glTF's JSON, if present

Depending on the solution, it should be opt-in or opt-out.

hth

**Comment by LoicFr** [comment]
Hi,

Thanks for the fix, it works and has a less performance impact than Blocovin's fix.
This also explains the LODGroup crash I encounter since it certainly also uses mesh bounds for computation.
